### DISCUSSION TITLE
EVP_CipherFinal_ex() faild

### DISCUSSION THREAD

**Discussion by so1616** [openssl/openssl#24924] [Q&A]
Hello.

I'm using OpenSSL3.0.14.
EVP_CipherFinal_ex() fails, which was successful in OpenSSL1.1.1.
The following error was output to the error log.
> 586C0000:error:1C80006B:Provider routines:ossl_cipher_generic_block_final:wrong final block length:providers\implementations\ciphers\ciphercommon.c:420:

Am I using EVP_CipherFinal_ex() or EVP_CIPHER_fetch() incorrectly?
I would like to know how I can encrypt it with AES.
Any advice would be much appreciated.

```
int encrypt(const char* in, int in_size, std::string* out)
{
	int nRet = 0;
	int nUpdateLen = 0, nFinalLen = 0;
	unsigned char aesIV[AES128_IV_LEN] = { 0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d,0x4d };
	unsigned char * tmp_pEncrypted = 0;
	
	EVP_CIPHER_CTX *ctx = lp_dll_EVP_CIPHER_CTX_new();
	/* <- Add For OpenSSL3.0 */
	EVP_CIPHER* cipher = lp_dll_EVP_CIPHER_fetch(NULL, "AES-128-CBC", NULL);
	if (cipher == NULL) {
		return -1;
	}
	/* -> */

	*out = "";

	BIO* errs = lp_dll_BIO_new_file("C:\\error.log", "w");
	
	do {
		tmp_pEncrypted = (unsigned char *)malloc(sizeof(unsigned char) * (in_size + AES128_IV_LEN) * 2);
		if (tmp_pEncrypted == 0)break;
		memset(tmp_pEncrypted, '\0', (in_size + AES128_IV_LEN) * 2);

		nUpdateLen = (int)(in_size)+EVP_MAX_BLOCK_LENGTH;
		lp_dll_EVP_CIPHER_CTX_reset(ctx);
		if (lp_dll_EVP_CIPHER_CTX_set_padding(ctx, 0) < 0) { nRet = -1; break; }
		/* <- OpenSSL1.1.1 */
		//if (!lp_dll_EVP_CipherInit_ex(ctx, lp_dll_EVP_aes_128_cbc(), NULL, pExtKey, aesIV, 1)) { nRet = -1; break; }
		/* -> */
		if (!lp_dll_EVP_CipherInit_ex(ctx, cipher, NULL, pExtKey, aesIV, 1)) { nRet = -1; break; }
		if (!lp_dll_EVP_CipherUpdate(ctx, tmp_pEncrypted, &nUpdateLen, (unsigned char *)in, (int)(in_size))) { nRet = -1; break; }
		int ret = lp_dll_EVP_CipherFinal_ex(ctx, tmp_pEncrypted + nUpdateLen, &nFinalLen);
		if (!ret) {
			lp_dll_ERR_print_errors(errs);
			lp_dll_EVP_CIPHER_free(cipher);
			return -1;
		}
		nFinalLen = (unsigned int)(nUpdateLen + nFinalLen);
		
		...
```

**Comment by t8m** [comment]
Try to change the order of the set_padding and CipherInit calls, that might help.

**Reply by so1616** [reply]
@t8m 
Thank you for your advice.

I tried the following order, but EVP_CipherFinal_ex() failed.
> EVP_CipherInit_ex()->EVP_CIPHER_CTX_set_padding()->EVP_CipherUpdate()
> EVP_CipherInit_ex()->EVP_CipherUpdate()->EVP_CIPHER_CTX_set_padding()

If I comment out EVP_CIPHER_CTX_set_padding(), EVP_CipherFinal_ex() succeeds, but decryption of the encrypted file fails.

**Comment by t8m** [comment]
So your data length is not a multiple of block size?

Then you actually cannot switch off padding with AES-128-CBC.

The resulting encrypted data will be at most one AES block larger than the plaintext data.

Not sure why do you allocate double of the input size although that should not matter.

**Reply by so1616** [reply]
in_size is 1380 bytes.
Am I correct in my understanding that in and in_size must be integer multiples of 16?
In that case, why did OpenSSL1.1.1 complete successfully without padding?

When padding is enabled, CTX_set_padding() returns 0.
 Is the ctx value incorrect?
`if (lp_dll_EVP_CIPHER_CTX_set_padding(ctx, 1) < 0) { nRet = -1; break; }`
In OpenSSL1.1.1, CTX_set_padding() always returns 1, but I found that this has changed in OpenSSL3.
I don't understand how the process has changed.

This is my first time dealing with encryption, so I'm sorry I don't fully understand it.

**Reply by so1616** [reply]
When I enabled CTX_set_padding(),  behavior was the same as OpenSSL1.1.1.
Thank you very much.
My understanding of encryption has deepened by studying based on the your advice on block size.
I appreciate it very much.

**Reply by t8m** [reply]
EVP_CIPHER_CTX_set_padding must be called after the EVP_CipherInit call where you set the algorithm (EVP_CIPHER). Before that, the context is not fully initialized.

**Reply by so1616** [reply]
Thank you for pointing it out.
Fixed my code.
